pkg/ingester/instance.go

```diff
@@ -381,27 +381,48 @@ func (i *instance) QuerySample(ctx context.Context, req logql.SelectSampleParams
 	return iters, nil
 }
 
-func (i *instance) Label(_ context.Context, req *logproto.LabelRequest) (*logproto.LabelResponse, error) {
-	var labels []string
-	if req.Values {
-		values, err := i.index.LabelValues(req.Name, nil)
-		if err != nil {
-			return nil, err
-		}
-		labels = make([]string, len(values))
-		for i := 0; i < len(values); i++ {
-			labels[i] = values[i]
+func (i *instance) Label(_ context.Context, req *logproto.LabelRequest, matchers ...*labels.Matcher) (*logproto.LabelResponse, error) {
+	if len(matchers) == 0 {
+		var labels []string
+		if req.Values {
+			values, err := i.index.LabelValues(req.Name, nil)
+			if err != nil {
+				return nil, err
+			}
+			labels = make([]string, len(values))
+			for i := 0; i < len(values); i++ {
+				labels[i] = values[i]
+			}
+		} else {
+			names, err := i.index.LabelNames(nil)
+			if err != nil {
+				return nil, err
+			}
+			labels = make([]string, len(names))
+			for i := 0; i < len(names); i++ {
+				labels[i] = names[i]
+			}
 		}
-	} else {
-		names, err := i.index.LabelNames(nil)
-		if err != nil {
-			return nil, err
+		return &logproto.LabelResponse{
+			Values: labels,
+		}, nil
+	}
+
+	ids, _ := i.index.Lookup(matchers, nil)
```

Comment:
If there's a matcher parameter it seems that we always return label values ?

Is that expected ? What if I pass a matcher and want label name ? Is this not supported ?


At the same time could we add documentation for that function. It wasn't there but since we're changing the behaviour it would be great to explain how matcher work.

Reply:
The Label method has been updated to handle a request for label names as well. A doc string was added as well.